chall04/migferna.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python3
+
+import sys
+import collections
+import os
+
+def order(bookshelfs, books):
+    bookshelfs = [int(x) for x in bookshelfs]
+    bookshelfs.sort(reverse=True)
+    books = sorted(books, key = lambda i: i['width'], reverse=True)
+    
+    i = 0
+    for book in books:
+        for i, bookshelf in enumerate(bookshelfs):
+            if book['width'] <= bookshelf:
+                bookshelf -= book['width']
+                bookshelfs[i] = bookshelf
+                break
+            elif len(bookshelfs) == i + 1:
+                return
+            else:
+                bookshelf + 1
+    return i if len(bookshelfs) == 0 else i+1
+
+def usage(message):
+    if (len(sys.argv) > 1):
+        print("%s: %s: %s" % (sys.argv[0], sys.argv[1], message))
+    else:
+        print("%s: stdin: %s" % (sys.argv[0], message))
+
+def open_files(filename):
+    books = []
+    bookshelfs = []
+    if (filename == os.path.basename(__file__)):
+        return bookshelfs, books
+    try:
+        with open(filename, "r") as file:
+            for i, line in enumerate(file):
+                if i == 0:
+                    bookshelfs =  line.split()
+                    bookshelfs = [s for s in bookshelfs if s.isdigit()]
+                else:
+                    line = line.split(" ", 1)
+                    book = {
+                        "width": int(line[0]),
+                        "title": line[1].rstrip("\n")
+                    }
+                    books.append(book)
```

Comment:
I don't like this `if i == 0:`, why putting this in the loop if you know that you will always execute this on the first line?  
```suggestion
            bookshelfs = [int(s) for s in f.readline().split()]
            for i, line in enumerate(file):
                line = line.split(" ", 1)
                book = {
                    "width": int(line[0]),
                    "title": line[1].rstrip("\n")
                }
                books.append(book)
```
You could simplify it even more by using a comprehension
```suggestion
            bookshelfs = [int(s) for s in f.readline().split()]
            books = [{
                    "width": int(line.split(" ", 1)[0]),
                    "title": line.split(" ", 1)[1].rstrip("\n")
                } for line in file]
```